### PR TITLE
Basic e2e

### PR DESCRIPTION
**Pull request by TrevorBrunette**
Basic end-to-end testing of board state

### CODE REVIEW

File: src/test/java/pro/trevor/tankgame/e2e/EndToEndTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                .collect(Collectors.toSet());
+        Set<String> actualDeadTanks = tester.getBoard().gatherUnits(GenericTank.class).stream()
+                .filter(Attribute.DEAD::unsafeFrom)
+                .map((t) -> t.getPlayerRef().getName())
+                .collect(Collectors.toSet());
+
+        assertEquals(actualLivingTanks, livingTanks);
+        assertEquals(actualDeadTanks, deadTanks);
+
+        noTanks.forEach((p) -> assertFalse(tester.getBoard().gatherUnits(GenericTank.class)
```
Comment (ryan3r) [inline src/test/java/pro/trevor/tankgame/e2e/EndToEndTestUtils.java:49]: I don't think there is a case where this catches anything.  Between actualLivingTanks and actualDeadTanks one of them should complain that there's an extra tank on the board.

File: src/test/java/pro/trevor/tankgame/e2e/EndToEndTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                .stream().map((t) -> t.getPlayerRef().getName())
+                .anyMatch(p::equals)));
+    }
+
+    public static <T> void assertPlayerAttributeEquals(EndToEndTester tester, String player, Attribute<T> attribute, T value) {
```
Comment (ryan3r) [inline src/test/java/pro/trevor/tankgame/e2e/EndToEndTestUtils.java:54]:
```suggestion
    public static <T> void assertPlayerElementAttributeEquals(EndToEndTester tester, String player, Attribute<T> attribute, T value) {
```
This checks the player element's attributes not the actual player attributes

File: src/test/java/pro/trevor/tankgame/e2e/TestEndToEndV3.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Test
+    public void testStateAtEndState() {
+        assertFalse(Attribute.RUNNING.unsafeFrom(tester.getState()));
+        assertEquals("Corey", Attribute.WINNER.unsafeFrom(tester.getState()));
+        assertEquals(18, Attribute.TICK.unsafeFrom(tester.getState()));
+        assertEquals(15, tester.getCouncil().allPlayersOnCouncil().size());
+        assertEquals(20, Attribute.COFFER.unsafeFrom(tester.getCouncil()));
+
+        tester.getCouncil().getCouncillors().forEach((p) -> assertFalse(tester.getCouncil().isPlayerSenator(p)));
```
Comment (ryan3r) [inline src/test/java/pro/trevor/tankgame/e2e/TestEndToEndV3.java:31]: These seems like a generic checks that would happen on most versions.  Helper function?

**Review by ryan3r — CHANGES_REQUESTED** [review]
Test cases not present that would be good to add at somepoint

* Gold mine locations (did they grow correctly)
* Empty tiles around gold mine (didn't grow too far)
* Wall health/locations

**Comment by TrevorBrunette** [comment]
> Test cases not present that would be good to add at somepoint
> 
> * Gold mine locations (did they grow correctly)
> * Empty tiles around gold mine (didn't grow too far)
> * Wall health/locations

Added.

**Review by ryan3r — APPROVED** [review]

**Pull request merged**